dashboard/public/frequency/frequency.js

```diff
@@ -0,0 +1,879 @@
+    .orient("left")
+    .ticks(5, "%");
+
+  /** @type {D3.selection} */
+  this.svg = this.container.append("svg").attr({
+      "width": this.getWidth() + this.margin.left + this.margin.right,
+      "height": this.getHeight() + this.margin.top + this.margin.bottom
+    }).append("g")
+    .attr("transform", "translate(" + this.margin.left + "," + this.margin.top + ")");
+
+  /** @type {D3.behavior} */
+  this.drag = this.createDragBehavior();
+
+  this.buildSVG();
+};
+
+/**
+ * Getter that calculates the height of the graph, taking into acount
+ * margins for UI elements.
+ * @returns {number} The height of the graph
+ */
+BarGraph.prototype.getHeight = function () {
+  return this.container.property("offsetHeight") - this.margin.top - this.margin.bottom;
+};
+
+/**
+ * Getter that calculates the width of the graph, taking into acount
+ * margins for UI elements.
+ * @returns {number} The width of the graph
+ */
+BarGraph.prototype.getWidth = function () {
+  return this.container.property("offsetWidth") - this.margin.left - this.margin.right;
+};
+
+/**
+ * @typedef {Object} ZippedData
+ * @property {user} UserData
+ * @property {english} Englishdata
+ */
+
+/**
+ * Getter that combines the two authoritative data sources
+ * (this.user_data and this.english_data) into a single list. Used to
+ * render the bars.
+ * @returns {Array.ZippedData}
+ */
+BarGraph.prototype.getZippedData = function (override) {
+  var user_data = override || this.user_data;
+  return user_data.map(function (_, i) {
+    return {
+      user: user_data[i],
+      english: this.english_data[i]
+    };
+  }, this);
+};
+
+/**
+ * Getter that examines both User and English data in the current order
+ * to return a mapping from User Data to English Data, with awareness of
+ * which points are locked and which are not.
+ * @returns {Object}  {
+ *                      user_letter: {
+ *                        letter: english_letter,
+ *                        locked: boolean
+ *                      },
+ *                      ...
+ *                    }
+ */
+BarGraph.prototype.getSubstitutionMap = function () {
+  return this.user_data.reduce(function (map, d, i) {
+    map[d.letter] = {
+      letter: this.english_data[i].letter,
+      locked: this.user_data[i].locked,
+    };
+    return map;
+  }.bind(this), {});
+};
+
+/**
+ * Helper method to process changes to user input. First updates the
+ * user frequency graph, then updates the output.
+ */
+BarGraph.prototype.processPlainText = function () {
+  this.updateUserDataFromInput();
+  this.processSubstitutions();
+};
+
+/**
+ * Method to populate the output based on the input and the user-defined
+ * substitutions
+ */
+BarGraph.prototype.processSubstitutions = function () {
+  var input = this.text_input.val();
+
+  /* this version preserves punctuation and special characters from the
+   * input in the output. If we want to strip them:
+   * var input = this.text_input.val().replace(/[^A-Za-z]/g, " ");
```

Comment:
Worth adding punctuation back to the sample text as well?  (Or choosing different sample text..)